src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs

```diff
@@ -427,6 +429,94 @@ public override bool ValidateChildren
             }
         }
 
+        /// <inheritdoc />
+        public override bool? HasValidators
+        {
+            get
+            {
+                if (!_hasValidators.HasValue)
+                {
+                    var visited = new HashSet<DefaultModelMetadata>();
+
+                    if (TryCalculateHasValidators(visited, this, out var hasValidators))
+                    {
+                        _hasValidators = hasValidators;
+                    }
+                    else
+                    {
+                        // Assume validation is required by default.
+                        _hasValidators = true;
+                    }
+                }
+
+                return _hasValidators.Value;
+            }
+        }
+
+        internal static bool TryCalculateHasValidators(HashSet<DefaultModelMetadata> visited, ModelMetadata metadata, out bool hasValidators)
+        {
+            RuntimeHelpers.EnsureSufficientExecutionStack();
+
+            if (metadata?.GetType() != typeof(DefaultModelMetadata))
+            {
+                // The calculation is valid only for DefaultModelMetadata instances. Null, other ModelMetadata instances
+                // or subtypes of DefaultModelMetadata will be treated as always requiring validation.
+                hasValidators = true;
+                return true;
+            }
+
+            var defaultModelMetadata = (DefaultModelMetadata)metadata;
+
+            if (defaultModelMetadata._hasValidators.HasValue)
+            {
+                // Return a previously calculated value if available.
+                hasValidators = defaultModelMetadata._hasValidators.Value;
+                return true;
+            }
+
+            if (defaultModelMetadata.ValidationMetadata.HasValidators != false)
+            {
+                // Either the ModelMetadata instance has some validators (HasValidators = true) or it is non-deterministic (HasValidators = null). 
+                // In either case, treat assume it has validators.
+                hasValidators = true;
+                return true;
+            }
+
+            // Before inspecting properties or elements of a collection, ensure we do not have a cycle.
+            // In case of a cycle, a ModelMetadata instance at a different (likely closer to the root) level
+            // is already querying if the metadata has validators and will eventually resolve a value.
+            // We'll no-op at this level and let the other instance calculate the value.
+            if (!visited.Add(defaultModelMetadata))
+            {
+                hasValidators = default;
```

Comment:
using default like this is *ehhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh*. I get it. but it would be easier to understand if it just said `false`.